COMPAT: Support for MPL 1.5

TomAugspurger opened the pull request:
closes https://github.com/pydata/pandas/issues/11111  (nice issue number)

Testing with tox locally I get 1 failure using either matplotlib 1.4 or 1.5. I may have messed up my tox config though. Still need to cleanup a bunch of stuff, I just threw in changes to get each test passing as I went.

release notes et. al coming tonight or tomorrow. Are we making any changes to the build matrix to run a matplotlib 1.5?

TomAugspurger committed:
COMPAT: Support for matplotlib 1.5

jreback commented:
u can add to the channel list (I think I showed this above)

it will thrn pick up the latest version if it's not specified exactly

jreback commented on pandas/tests/test_graphics.py:
```diff
@@ -74,13 +74,23 @@ def setUp(self):
         else:
             self.bp_n_objects = 8
+        # Thier dev version is messed up, 0+unknown
+        if mpl_version[0] == '0' or mpl_version >= LooseVersion('1.5'):
```
prob best to put these in some functions (either in ``util.testing`` or top-level here)

jreback commented on pandas/tests/test_graphics.py:
```diff
@@ -183,7 +193,7 @@ def _check_colors(self, collections, linecolors=None, facecolors=None,
 
         from matplotlib.lines import Line2D
-        from matplotlib.collections import Collection
+        from matplotlib.collections import Collection, PolyCollection
```
is this always availabel?

TomAugspurger replied:
at leas as of 1.2.0, which I think is our oldest supported version.

jreback replied:
@TomAugspurger as an aside
if it makes things easier now (or in a later version )
we could move up the min supported version of matplotlib from 1.2
cc @tacaswell

tacaswell replied:
For reference, mpl does not support past the current minor release (ex, we currently do not fix bugs in 1.3.x, and in a short while will not fix bugs in 1.4.x).

jreback commented on pandas/tools/plotting.py:
```diff
@@ -31,6 +31,16 @@
+    import matplotlib
+    mpl_v = matplotlib.__version__
+    if mpl_v >= LooseVersion('1.5') or mpl_v[0] == '0':
+        mpl_ge_1_5_0 = True
```
definitly should be a function somewhere

TomAugspurger committed:
COMPAT/BUG: color handling in scatter

MPL 1.5 seems to have changed the behavior of

ax.scatter(x, y, c='green', color=['blue'])

Since this doesn't really make sense anyway(?) I'm changing
our code so that the user doesn't get into that state.

TomAugspurger committed:
TST: Capture warnings in _check_plot_works

TomAugspurger commented on pandas/core/config_init.py:
```diff
@@ -214,7 +214,8 @@
     import sys
-    from pandas.tools.plotting import mpl_stylesheet
+    from pandas.tools.plotting import mpl_stylesheet, _update_stylesheet
+    _update_stylesheet()
```
This is pretty hacky, but `mpl_stylesheet` is a dict defined in pandas.tools.plotting. I have my matplotlib version check in `pandas.util.testing`, but I can't run code from that in the main body of `pandas.tools.plotting` (I'm assuming a circular import).

TomAugspurger replied:
Bah, this doesn't work since you can't set the option multiple times.

jreback commented on pandas/tools/plotting.py:
```diff
@@ -24,13 +24,18 @@
     pass
 
+def _mpl_ge_1_5_0():
+    import matplotlib
```
seems you already have this in ``utils.testing``?

jreback commented on pandas/tseries/tests/test_plotting.py:
```diff
@@ -486,9 +486,9 @@ def test_gaps(self):
-        mask = data.mask
-        self.assertTrue(mask[5:25, 1].all())
+        # tm.assertIsInstance(data, np.ma.core.MaskedArray)
+        # mask = data.mask
```
these just don't work? put skips rather than comment out then

jreback commented on pandas/util/testing.py:
```diff
@@ -41,6 +41,7 @@
 from pandas.util.decorators import deprecate
 from pandas import _testing
 from pandas.io.common import urlopen
+from pandas.tools.plotting import _mpl_ge_1_5_0
```
don't do this, just define it here

jreback commented:
is this not tested on travis? I don't see the icon

tacaswell commented on ci/install_conda.sh:
```diff
@@ -72,6 +72,7 @@ bash miniconda.sh -b -p $HOME/miniconda || exit 1
 
 conda config --set always_yes yes --set changeps1 no || exit 1
 conda update -q conda || exit 1
+conda config --add channels tacaswell || exit 1
```
the conda-forge channel might be better to use here

TomAugspurger commented:
Travis was flaky yesterday morning. I just pushed some changes, and it's picked up now.

TomAugspurger commented:
@jreback travis is green. Looks like https://travis-ci.org/pydata/pandas/jobs/81429896 and https://travis-ci.org/pydata/pandas/jobs/81429897 both ran with matplotlib RC1.

jreback commented:
make the 3.4 regular build not specify a matplotlib (in requirements-3.4.run). All of the slow tests are running (e.g. 2.7) just fine. But I don't think the fast ones actually are run by travis (in theory they are on 3.3, but that is not build ATM for 1.5.0rc1).

TomAugspurger commented:
@jreback done. We'll want to followup and pin it at 1.5 once that's released. I'm assuming the `conda-forge` channel will continue to track master and we won't want changes there breaking our builds.

jreback commented:
thanks @TomAugspurger

tacaswell commented:
@TomAugspurger I don't think we plan to update conda forge from master, just tags.
